HDDS-5201. [FSO] S3MultiPart: Use existing ozone key format for MPU Info in DB

rakeshadr opened the pull request:
## What changes were proposed in this pull request?

This task is to simplify the prefix layout implementation for the S3MPU keys. As there is no ObjectStore API to perform rename or delete operation on MPU keys, the DB table key format can follow the existing structure for the MPU table entries.

This will help to reuse the existing ozoneBucket#listParts() and ozoneBucket#listMultipartUploads() and required very minimal code changes.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5201

## How was this patch tested?

Added and modify unit & integration test cases.

rakeshadr commented:
**Note:** Since I have touched all the S3 flows - initiate -> commit -> complete and abort the patch has modified all these source and test classes.

linyiqun reviewed (commented):
Haven't taken a detailed review,  some initial comments below:

mukul1987 commented:
i do not think, if we have a bucket rename API available right now.

However with HDDS-2939, even if the bucket it renamed, its object id will still remain the same, so we will still be to use the code

rakeshadr commented:
> Haven't taken a detailed review, some initial comments below.
> In addition, I have a question for this, @rakeshadr As you mentioned, MPU api cannot support renamed. But if there will be a corner case that S3 bucket can be renamed by admin user since currently we use specific bucket as the S3 bucket.
> After S3 bucket be renamed, will current MPU logic be broken?

Thanks @linyiqun for the review comments. 
I'm adding @mukul1987 's comment link [here](https://github.com/apache/ozone/pull/2228/#issuecomment-838811615) to chain the discussion thread..

linyiqun commented:
>However with HDDS-2939, even if the bucket it renamed, its object id will still remain the same, so we will still be to use the code

Yes, the object id stored in file/dir table can make sense here. But multipartInfoTable didn't stored the object id, it stores the full path as original logic. Maybe this is not a big problem now, I just bring up this corner case that I can Imagine.

rakeshadr commented:
> Yes, the object id stored in file/dir table can make sense here. But multipartInfoTable didn't stored the object id, it stores the full path as original logic. Maybe this is not a big problem now, I just bring up this corner case that I can Imagine.

@linyiqun Thanks for adding more details and its really a good thought. IMHO, since Ozone don't support bucket#rename API this case won't hit now. Also, I searched bucket#rename API in S3. I could see that they are not providing bucket#rename API in S3 and recommended users to do copy keys to a new bucket.

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/multipart/S3MultipartUploadAbortRequestWithFSO.java:
```diff
@@ -54,19 +54,20 @@ protected OMClientResponse getOmClientResponse(IOException exception,
-  protected String getMultipartKey(String multipartUploadID, String volumeName,
-      String bucketName, String keyName, OMMetadataManager omMetadataManager)
-      throws IOException {
+  protected String getMultipartOpenKey(String multipartUploadID,
```
Done!

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/multipart/S3MultipartUploadCommitPartRequestWithFSO.java:
```diff
@@ -62,197 +46,41 @@ public S3MultipartUploadCommitPartRequestWithFSO(OMRequest omRequest) {
 
-    return omClientResponse;
+  @SuppressWarnings("checkstyle:ParameterNumber")
+  protected S3MultipartUploadCommitPartResponse getOmClientResponse(
```
Done!

rakeshadr commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/s3/multipart/TestS3MultipartUploadCommitPartRequestWithFSO.java:
```diff
@@ -66,13 +66,20 @@ protected void addKeyToOpenKeyTable(String volumeName, String bucketName,
   }
 
-  protected String getMultipartKey(String volumeName, String bucketName,
+  protected String getMultipartOpenKey(String volumeName, String bucketName,
```
Thanks @linyiqun for the comment. I have visited all the FSO files and I hope all the cases has been taken care now.

linyiqun commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneClientMultipartUploadWithFSO.java:
```diff
@@ -846,8 +930,7 @@ private String verifyUploadedPart(String volumeName, String bucketName,
 
-    long parentID = getParentID(volumeName, bucketName, keyName,
-        metadataMgr);
+    getParentID(volumeName, bucketName, keyName, metadataMgr);
```
Nit: This line can be removed.

rakeshadr replied:
Done!

linyiqun reviewed (commented):
@rakeshadr , thanks for addressing the review comment. Only catch one nit after more deep reviewed. 
BTW, one latest CI check is failed. Please have a look.

rakeshadr commented:
> @rakeshadr , thanks for addressing the review comment. Only catch one nit after more deep reviewed.
> BTW, one latest CI check is failed. Please have a look.

Thanks again @linyiqun for the reviews. The tests failures are unrelated. Multi-delete and s3, acceptance test are failing, these are running with simple layout version and are random failures. I've pushed one more commit and will monitor next CI run.

linyiqun reviewed (approved):
LGTM, +1.

rakeshadr commented:
Thanks @linyiqun for the reviews.
Thanks @mukul1987 , @bharatviswa504 , @sadanand48 for the offline discussions.

I've merged the changes to the feature branch and with this hopefully s3 acceptance test would be able to run happily.
